### PR TITLE
glib: switch elfutils in for abandoned libelf

### PR DESCRIPTION
**Pull request by philiptaron**
## Description of changes

Swapped out `libelf` for `elfutils` in `glib`; as #271473 notes:

> `libelf` did not see an update since 2014: https://github.com/Distrotech/libelf. `elfutils` is a popular API-compatible ELF parser that most users expect.

I targeted staging since `glib` is at the base of a lot of packages. I did build various GNOME things (`gnome.gnome-shell`, for instance). 

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by philiptaron** [comment]
It appears that there are [issues on macOS](https://gist.github.com/GrahamcOfBorg/bdc734b7259c1d1a284a5166a0c295ce).

File: pkgs/development/libraries/glib/default.nix [hunk 1]
```diff
@@ -151,6 +151,8 @@ stdenv.mkDerivation (finalAttrs: {
     "-Dgtk_doc=${lib.boolToString buildDocs}"
     "-Dnls=enabled"
     "-Ddevbindir=${placeholder "dev"}/bin"
+  ] ++ lib.optionals (!stdenv.isLinux) [
+    "-Dlibelf=disabled"
```
Comment (philiptaron) [inline pkgs/development/libraries/glib/default.nix:155]: This controls a few ifdefs in [gio/gresource-tool.c](https://gitlab.gnome.org/GNOME/glib/-/blob/main/gio/gresource-tool.c) only.

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by philiptaron** [comment]
Rebased on latest staging to avoid evail failed due to tzdata bug https://github.com/NixOS/nixpkgs/pull/292480.

**Comment by SuperSandro2000** [comment]
Not sure what fails here and why

```
I/O error : Attempt to load network entity http://www.oasis-open.org/docbook/xml/4.2/docbookx.dtd
/build/glib-2.78.4/docs/reference/gio/gio.xml:2: warning: failed to load external entity "http://www.oasis-open.org/docbook/xml/4.2/docbookx.dtd"
D DocBook XML V4.2//EN" "http://www.oasis-open.org/docbook/xml/4.2/docbookx.dtd"
                                                                               ^
I/O error : Attempt to load network entity http://www.oasis-open.org/docbook/xml/4.1.2/docbookx.dtd
../../../../gio/tests/gdbus-object-manager-example/objectmanager-gen-org.gtk.GDBus.Example.ObjectManager.Animal.xml:4: warning: failed to load external entity "http://www.oasis-open.org/docbook/xml/4.1.2/docbookx.dtd"
]>
  ^
I/O error : Attempt to load network entity http://www.oasis-open.org/docbook/xml/4.1.2/docbookx.dtd
../../../../gio/tests/gdbus-object-manager-example/objectmanager-gen-org.gtk.GDBus.Example.ObjectManager.Cat.xml:4: warning: failed to load external entity "http://www.oasis-open.org/docbook/xml/4.1.2/docbookx.dtd"
]>
  ^
Package glib-2.0 was not found in the pkg-config search path.
Perhaps you should add the directory containing `glib-2.0.pc'
to the PKG_CONFIG_PATH environment variable
No package 'glib-2.0' found
```

**Comment by philiptaron** [comment]
Fix for the test failure is in #292687.

**Comment by SuperSandro2000** [comment]
@ofborg build glib

**Review by SuperSandro2000 — APPROVED** [review]

File: pkgs/development/libraries/glib/default.nix (hunk 1 above)
Comment (jtojnar) [reply pkgs/development/libraries/glib/default.nix:155]: The feature allows dumping gresources embedded into ELF files (e.g. `gresource list result/lib/libgtk-3.so.0`). Not sure how important that is on Darwin when they use Mach-O instead, probably only if someone is debugging Linux programs on Mac.

**Comment by philiptaron** [comment]
@SuperSandro2000, was there anything else you are waiting on me for before merging?

**Comment by SuperSandro2000** [comment]
One test is failing :/

``` 
 93/341 glib:glib+core / autoptr                                                        OK               0.13s   56 subtests passed
 94/341 glib:glib+core / regex                                                          ERROR            0.65s   killed by signal 6 SIGABRT
>>> MALLOC_CHECK_=2 MALLOC_PERTURB_=203 LD_LIBRARY_PATH=/build/glib-2.78.4/build/glib:/build/glib-2.78.4/glib/.libs G_TEST_SRCDIR=/build/glib-2.78.4/glib/tests G_ENABLE_DIAGNOSTIC=1 ASAN_OPTIONS=halt_on_error=1:abort_on_error=1:print_summary=1 G_DEBUG=gc-friendly G_TEST_BUILDDIR=/build/glib-2.78.4/build/glib/tests UBSAN_OPTIONS=halt_on_error=1:abort_on_error=1:print_summary=1:print_stacktrace=1 /build/glib-2.78.4/build/glib/tests/regex
 95/341 glib:glib+core / shell                                                          OK               0.62s   51 subtests passed
 96/341 glib:glib+core / include                                                        OK               0.36s   1 subtests passed
 ```

**Comment by philiptaron** [comment]
> One test is failing :/
> 
> ```
>  93/341 glib:glib+core / autoptr                                                        OK               0.13s   56 subtests passed
>  94/341 glib:glib+core / regex                                                          ERROR            0.65s   killed by signal 6 SIGABRT
> >>> MALLOC_CHECK_=2 MALLOC_PERTURB_=203 LD_LIBRARY_PATH=/build/glib-2.78.4/build/glib:/build/glib-2.78.4/glib/.libs G_TEST_SRCDIR=/build/glib-2.78.4/glib/tests G_ENABLE_DIAGNOSTIC=1 ASAN_OPTIONS=halt_on_error=1:abort_on_error=1:print_summary=1 G_DEBUG=gc-friendly G_TEST_BUILDDIR=/build/glib-2.78.4/build/glib/tests UBSAN_OPTIONS=halt_on_error=1:abort_on_error=1:print_summary=1:print_stacktrace=1 /build/glib-2.78.4/build/glib/tests/regex
>  95/341 glib:glib+core / shell                                                          OK               0.62s   51 subtests passed
>  96/341 glib:glib+core / include                                                        OK               0.36s   1 subtests passed
> ```

Does ofBorg not rebase before testing? 😮‍💨 

That's the test that's fixed in https://github.com/NixOS/nixpkgs/pull/292687. 

On a local rebased version I build successfully.

```
┏━ Dependency Graph with 2 roots:
┃ ┌─ ✔ desktop-file-utils-0.27 ⏱ 2s
┃ ├─ ✔ shared-mime-info-2.4 ⏱ 9s
┃ ✔ glib-2.78.4 ⏱ 5m13s
┃          ┌─ ✔ libmicrohttpd-0.9.77 ⏱ 42s
┃       ┌─ ✔ elfutils-0.190 ⏱ 1m13s
┃       │              ┌─ ✔ python3.11-execnet-2.0.2 ⏱ 6s
┃       │              │     ┌─ ✔ git-2.43.2 ⏱ 2m39s
┃       │              │  ┌─ ✔ python3.11-hatch-vcs-0.4.0 ⏱ 2s
┃       │              ├─ ✔ python3.11-filelock-3.13.1 ⏱ 9s
┃       │           ┌─ ✔ python3.11-pytest-xdist-3.5.0 ⏱ 8s
┃       │        ┌─ ✔ python3.11-requests-2.31.0 ⏱ 14s
┃       │     ┌─ ✔ python3.11-sphinx-7.2.6 ⏱ 2m43s
┃       │  ┌─ ✔ python3.11-pip-23.3.1 ⏱ 12s
┃       ├─ ✔ libselinux-3.6 ⏱ 20s
┃       ├─ ⏸ perl-5.38.2
┃    ┌─ ✔ glib-2.78.4 ⏱ 9m50s
┃    │  ┌─ ⏸ pkg-config-0.29.2
┃    ├─ ⏸ pkg-config-wrapper-0.29.2
┃ ┌─ ✔ check-pkg-config-gio-2.0-gobject-2.0-gthread-2.0
┃ ✔ check-meta-pkg-config-modules-for-glib-2.78.4 ⏱ 5m23s
┣━━━ Builds            │ Downloads       │ Host
┃        │ ✔ 461 │     │     │     │     │ localhost
┃        │       │     │     │ ↓ 2 │     │ https://cache.nixos.org
┗━ ∑ ⏵ 0 │ ✔ 461 │ ⏸ 5 │ ↓ 0 │ ↓ 2 │ ⏸ 0 │ Finished at 14:31:18 after 1h28m27s
/nix/store/6vzralsyq68i06jvm1g6b7wxyvy0janx-glib-2.78.4-bin
/nix/store/pgycahl8ws9pyq71iv48m9mxz0sw59g6-check-meta-pkg-config-modules-for-glib-2.78.4
/nix/store/wbi3ad6nxb1hpskhdk0jc7xfv71h9im6-glib-2.78.4-bin
```

**Comment by philiptaron** [comment]
Thank you for being careful @SuperSandro2000. I appreciate you.

**Comment by alyssais** [comment]
Broke static; fix is https://github.com/NixOS/nixpkgs/pull/298177.

**Pull request merged**